File: src/recipejar/domain/Recipe.java
package recipejar.domain;

import java.util.ArrayList;
import java.util.List;
import recipejar.recipe.Ingredient;
import recipejar.util.LabelUtils;

/**
 * Simple domain model for a recipe (data only, no UI or persistence concerns).
 * Per plan Phase 3.
 */
public class Recipe {

    private String title = "";
    private String notes = "";
    private String procedure = "";
    private List<Ingredient> ingredients = new ArrayList<>();
    private List<String> labels = new ArrayList<>();

    public Recipe() {}

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title != null ? title : ""; }

    public String getNotes() { return notes; }
    public void setNotes(String notes) { this.notes = notes != null ? notes : ""; }

    public String getProcedure() { return procedure; }
    public void setProcedure(String procedure) { this.procedure = procedure != null ? procedure : ""; }

    public List<Ingredient> getIngredients() { return ingredients; }
    public void setIngredients(List<Ingredient> ingredients) {
        this.ingredients = ingredients != null ? ingredients : new ArrayList<>();
    }

    public List<String> getLabels() { return labels; }
    public void setLabels(List<String> labels) {
        this.labels = labels != null ? labels : new ArrayList<>();
    }

    public void addLabel(String label) {
        if (label != null && !label.trim().isEmpty() && !labels.contains(label.trim())) {
            labels.add(label.trim());
        }
    }

    public boolean isLabeled(String label) {
        for (String existing : labels) {
            if (LabelUtils.matches(existing, label)) {
                return true;
            }
        }
        return false;
    }
}

File: shared/src/commonMain/kotlin/recipejar/domain/Recipe.kt
package recipejar.domain

import recipejar.recipe.Ingredient

/**
 * Core domain model, ported/adapted from original Recipe/RecipeFile data.
 * Focus on fields needed for HTML roundtrip: title, notes, ingredients, procedure, labels, meta.
 * No UI/docs listeners.
 * NOTE: ingredients/labels/meta are mutable for parse/serialize efficiency (per port of orig ArrayList/HashMap); do not share instances across recipes.
 */
data class Recipe(
    var title: String = "",
    var notes: String = "",
    val ingredients: MutableList<Ingredient> = mutableListOf(),
    var procedure: String = "",
    val labels: MutableList<String> = mutableListOf(),
    val meta: MutableMap<String, String> = mutableMapOf()
) {
    fun setLabels(text: String) {
        labels.clear()
        if (text.isNotBlank()) {
            labels.addAll(text.split(",").map { it.trim() }.filter { it.isNotEmpty() })
        }
    }

    fun getLabelsAsString(): String = labels.joinToString(", ")

    fun addLabel(s: String) {
        if (s.isNotBlank() && !labels.contains(s.trim())) labels.add(s.trim())
    }
}
